### PR TITLE
f/cnn

### PR DESCRIPTION
**Pull request by whittlbc**
Adding first version of CNN with script to train it as well

### CODE REVIEW

File: format_input_data.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      
+      info['labels'].append(dml)
+      
+      image_as_array = misc.imread(image_path, mode='RGB')
+      
+      info['data'].append(image_as_array)
+      
+      i += 1
+    
+    info['labels'] = np.array(info['labels'])  # Do these labels need to be converted to another datatype?
```
Comment (whittlbc) [inline format_input_data.py]: ?

File: train_cnn.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+Image (625x1250)
+  --> 3x3 receptive fields convolved with stride 1 as in VGGNet (this done twice)
+    --> fixed-size output vector
+      --> CNN(width=32)
+        --> CNN(width=64)
+          --> CNN(width=128)
+            --> 2 fully-connected layers (size=1024) applying the rectified linear unit activation
+              --> output vector (to be concatenated with 1st LSTM output vector)
+
+"""
```
Comment (whittlbc) [inline train_cnn.py]: ^Description of their CNN

File: train_cnn.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+def get_batch(X, Y, N, n=32):
+  inds = np.random.choice(range(N), size=n, replace=False)
+  return X[inds], Y[inds]
+
+
+if __name__ == '__main__':
+  # Extract pickled train, validation, and test data
+  X_train, Y_train = extract_data('{}/train.pkl'.format(data_dir))
+  X_val, Y_val = extract_data('{}/validation.pkl'.format(data_dir))
+  X_test, Y_test = extract_data('{}/test.pkl'.format(data_dir))
```
Comment (whittlbc) [inline train_cnn.py]: ^Using a train/validation/test ratio of 0.6/0.2/0.2 right now, with only 10 total images just to get the hang of things.

File: train_cnn.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  
+  # Define our params
+  num_train_images = X_train.shape[0]  # => 6
+  train_steps = num_train_images
+  N = num_train_images
+  Nbatch = 1
+  print_step = 6
+  W = X_train.shape[1]  # => 1250
+  H = W
+  C = 3  # What is C, and why 3?
```
Comment (whittlbc) [inline train_cnn.py]: ?

File: train_cnn.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  filters = 4  # Why?
```
Comment (whittlbc) [inline train_cnn.py]: Why 4? Also, what exactly does this signify?

File: train_cnn.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  dims = [3, 3]  # is this the size of sliding window?
```
Comment (whittlbc) [inline train_cnn.py]: ?

File: train_cnn.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  strides = [1, 1]
+  act = tf.nn.relu
+  std = 1e-2
+  num_classes = num_train_images  # each image should have its own unique label, so num_classes should equal num images?
```
Comment (whittlbc) [inline train_cnn.py]: ?

File: train_cnn.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  init = tf.global_variables_initializer()
+  
+  train_hist = []
+  val_hist = []
+  
+  sess = tf.Session()
+  sess.run(init)
+  
+  # Why do this here before going into for-loop?
+  xb, yb = get_batch(X_train, Y_train, N, n=Nbatch)
```
Comment (whittlbc) [inline train_cnn.py]: ?

File: train_cnn.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      
+      train_hist.append(l)
+      val_hist.append(lval)
+
+  a = sess.run(o, {x: xb, y: yb})
+
+  # What are these?
+  print a[:2]
+  print a.shape
+  print a[0] / np.sum(a[0])
```
Comment (whittlbc) [inline train_cnn.py]: ^?

File: create_vocab.py [hunk 10]
```diff
@@ -1,6 +1,9 @@
+"""
```
Comment (whittlbc) [inline create_vocab.py:1]: Ignore file for now, Gab. Not CNN related.

File: glimpse/decoders/lstm.py [hunk 11]
```diff
@@ -0,0 +1,6 @@
+from glimpse.neural_net import NeuralNet
+
+
+class LSTM(NeuralNet):
```
Comment (whittlbc) [inline glimpse/decoders/lstm.py:4]: placeholder class for when I get to it

File: format_input_data.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
```
Comment (whittlbc) [inline format_input_data.py:73]: Do my labels need to be encoded from DML (a string that looks like `'<0 a="16px" b="0 auto"...'`) to an array of integers using some sort of `char_2_id` map, such as `{'<': 0, 'a': 1, 'b': 2, ...}`)?

Comment (whittlbc) [reply format_input_data.py:73]: Right now the numpy datatype is apparently `S8749`

File: glimpse/decoders/params.json [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (whittlbc) [inline glimpse/decoders/params.json:1]: placeholder params file for when I get to writing encoder LSTM

File: glimpse/encoders/cnn.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+    # Extract pickled train, validation, and test data
+    self.extract_data()
+    
+    # Define non-static params left out of params.json
+    self.define_dynamic_params()
+    
+    # Establish weight placeholders
+    self.x = tf.placeholder(shape=[None, self.params.W, self.params.H, self.params.C], dtype=tf.float32)
+    self.y = tf.placeholder(shape=[None], dtype=tf.int32)  # wrong datatype --> currently dealing with dtype='|S8749'
```
Comment (whittlbc) [inline glimpse/encoders/cnn.py:32]: See above comment inside `format_input_data.py`

File: glimpse/encoders/cnn.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+      
+      x, y = data['data'], data['labels']
+      
+      setattr(self, 'X_{}'.format(set), x)
+      setattr(self, 'Y_{}'.format(set), y)
+    
+    print 'X_train.shape = {}, Y_train.shape = {}'.format(self.X_train.shape, self.Y_train.shape)
+    
+  def define_dynamic_params(self):
+    num_train_images = self.X_train.shape[0]  # => 6 right now
```
Comment (whittlbc) [inline glimpse/encoders/cnn.py:51]: Using only 10 images right now just to get the hang of things in a 6/2/2 breakdown among train/val/test

File: glimpse/encoders/cnn.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+      'num_classes': num_train_images  # each image should have its own unique label... so this should equal num_images?
+    }
+    
+    [setattr(self.params, k, v) for k, v in dyn_params.iteritems()]
+    
+  def build_nn(self):
+    # Add 3 Conv2D layers
+    self.add_conv(n_layers=self.n_conv_layers)
+    
+    # Reshape o for some goddamn reason
```
Comment (whittlbc) [inline glimpse/encoders/cnn.py:70]: What we doing this for?

File: glimpse/encoders/cnn.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+        lval = sess.run(loss, {self.x: xb, self.y: yb})
+    
+        print "iter: {} Train: {}, Val: {}".format(i, l, lval)
+    
+        train_hist.append(l)
+        val_hist.append(lval)
+
+    a = sess.run(self.o, {self.x: xb, self.y: yb})
+
+    # What are these?
```
Comment (whittlbc) [inline glimpse/encoders/cnn.py:114]: ?

File: glimpse/encoders/lstm.py [hunk 18]
```diff
@@ -0,0 +1,6 @@
+from glimpse.neural_net import NeuralNet
+
+
+class LSTM(NeuralNet):
```
Comment (whittlbc) [inline glimpse/encoders/lstm.py:4]: placeholder

File: glimpse/encoders/params.json [hunk 19]
```diff
@@ -0,0 +1,17 @@
+{
+  "cnn": {
+    "Nbatch": 1,
+    "print_step": 6,
+    "C": 3,
```
Comment (whittlbc) [inline glimpse/encoders/params.json:5]: What is C and why 3?

File: glimpse/encoders/params.json [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+    "filters": 4,
```
Comment (whittlbc) [inline glimpse/encoders/params.json:6]: Why 4 filters and what does this represent?

File: glimpse/encoders/params.json [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+    "dims": [3, 3],
```
Comment (whittlbc) [inline glimpse/encoders/params.json:7]: Is this "sliding window" size in pixels?
